File: cpp/scsi.cpp
#include "scsi.h"
#include <malloc.h> // For _aligned_malloc
#include <stdexcept>
#include <intrin.h> // For byte swap

namespace scsi
{

    // --- ScsiDataBuffer implementation ---
    ScsiDataBuffer::ScsiDataBuffer(size_t size) : size_(size)
    {
        aligned_buffer_ = _aligned_malloc(size, SCSI_DATA_BUFFER_ALIGNMENT);
        if (!aligned_buffer_)
        {
            throw std::bad_alloc();
        }
    }

    ScsiDataBuffer::~ScsiDataBuffer()
    {
        if (aligned_buffer_)
        {
            _aligned_free(aligned_buffer_);
        }
    }

    uint8_t *ScsiDataBuffer::data()
    {
        return static_cast<uint8_t *>(aligned_buffer_);
    }

    const uint8_t *ScsiDataBuffer::data() const
    {
        return static_cast<const uint8_t *>(aligned_buffer_);
    }

    size_t ScsiDataBuffer::size() const
    {
        return size_;
    }

    uint8_t &ScsiDataBuffer::operator[](size_t index)
    {
        if (index >= size_)
        {
            throw std::out_of_range("Index out of range");
        }
        return data()[index];
    }

    const uint8_t &ScsiDataBuffer::operator[](size_t index) const
    {
        if (index >= size_)
        {
            throw std::out_of_range("Index out of range");
        }
        return data()[index];
    }

    // --- SCSI_PASS_THROUGH_DIRECT_WITH_BUFFER implementation ---
    SCSI_PASS_THROUGH_DIRECT_WITH_BUFFER::SCSI_PASS_THROUGH_DIRECT_WITH_BUFFER(int dir)
        : sptd(), Filler(0), ucSenseBuf()
    {
        sptd.Length = sizeof(SCSI_PASS_THROUGH_DIRECT);
        sptd.ScsiStatus = 0;
        sptd.PathId = 0;
        sptd.TargetId = 0;
        sptd.Lun = 0;
        sptd.CdbLength = 0; // Must be set by user
        sptd.DataIn = static_cast<UCHAR>(dir);
        sptd.DataBuffer = nullptr;
        sptd.DataTransferLength = 0;
        sptd.SenseInfoOffset = offsetof(SCSI_PASS_THROUGH_DIRECT_WITH_BUFFER, ucSenseBuf);
        sptd.SenseInfoLength = sizeof(SenseBuffer);
        sptd.TimeOutValue = 10;
        // Zero the CDB
        memset(sptd.Cdb, 0, sizeof(sptd.Cdb));
    }

    void SCSI_PASS_THROUGH_DIRECT_WITH_BUFFER::set_buffer(int dir, ScsiDataBuffer &data_src)
    {
        set_buffer(dir, data_src.data(), data_src.size());
    }

    void SCSI_PASS_THROUGH_DIRECT_WITH_BUFFER::set_buffer(int dir, void *data_ptr, size_t data_len)
    {
        sptd.DataIn = static_cast<UCHAR>(dir);
        sptd.DataBuffer = data_ptr;
        sptd.DataTransferLength = static_cast<ULONG>(data_len);
    }

    void SCSI_PASS_THROUGH_DIRECT_WITH_BUFFER::set_buffer(int dir, std::span<const uint8_t> data_src)
    {
        sptd.DataIn = static_cast<UCHAR>(dir);
        // The Windows API expects a non-const pointer. We are promising not to write to it when dir is SCSI_IOCTL_DATA_OUT.
        sptd.DataBuffer = const_cast<uint8_t *>(data_src.data());
        sptd.DataTransferLength = static_cast<ULONG>(data_src.size());
    }

    std::ostream &operator<<(std::ostream &os, const SCSI_PASS_THROUGH_DIRECT_WITH_BUFFER &s)
    {
        os << "ScsiPassThrough: ScsiStatus " << static_cast<int>(s.sptd.ScsiStatus)
           << ", DataTransferLength " << s.sptd.DataTransferLength
           << ", CDB: opcode " << static_cast<int>(s.sptd.Cdb[0])
           << ", flag " << static_cast<int>(s.sptd.Cdb[1]);
        return os;
    }

    // --- Byte swap implementation ---
    template <>
    uint16_t swap_bytes<uint16_t>(uint16_t value)
    {
        return _byteswap_ushort(value);
    }

    template <>
    uint32_t swap_bytes<uint32_t>(uint32_t value)
    {
        return _byteswap_ulong(value);
    }

    template <>
    uint64_t swap_bytes<uint64_t>(uint64_t value)
    {
        return _byteswap_uint64(value);
    }

    // --- CDB constructors ---
    ScsiRwCdb16::ScsiRwCdb16(ScsiOpcode op, uint64_t lba_in, uint32_t len_in, uint8_t flags_in)
        : opcode(static_cast<uint8_t>(op)), flags(flags_in), lba(swap_bytes(lba_in)) // Big Endian
          ,
          len(swap_bytes(len_in)) // Big Endian
          ,
          group(0), control(0)
    {
    }

    ScsiSecCdb12::ScsiSecCdb12(ScsiOpcode op, uint8_t protocol_in, uint16_t com_id_in, uint32_t len_in)
        : opcode(static_cast<uint8_t>(op)), protocol(protocol_in), com_id(swap_bytes(com_id_in)) // Big Endian
          ,
          reserved(0), len(swap_bytes(len_in)) // Big Endian
          ,
          reserved2(0), control(0)
    {
    }

} // namespace scsi

File: rust/src/dev/scsi.rs
use endian_codec::{DecodeBE, EncodeBE, PackedSize};
use memoffset::offset_of;
use std::{
    ffi::c_void,
    fmt,
    mem::size_of,
    ops::{Index, IndexMut},
    ptr::null_mut,
};
use windows_sys::Win32::Storage::IscsiDisc::SCSI_PASS_THROUGH_DIRECT;

// Shouldn't have more than 255 bytes
pub type SenseBuffer = [u8; 32];

// Alignment must match that of the device https://docs.microsoft.com/en-us/windows-hardware/drivers/ddi/ntddscsi/ni-ntddscsi-ioctl_scsi_pass_through_direct#remarks
// Since we don't know it we use the max alignment: double DWORD https://docs.microsoft.com/en-us/windows-hardware/drivers/ddi/ntddstor/ns-ntddstor-_storage_adapter_descriptor
#[repr(C, align(64))]
pub struct ScsiDataBuffer(Vec<u8>);

#[repr(C)]
#[allow(non_camel_case_types, non_snake_case, unused_assignments)]
pub struct SCSI_PASS_THROUGH_DIRECT_WITH_BUFFER {
    pub sptd: SCSI_PASS_THROUGH_DIRECT,
    pub Filler: u32,
    pub ucSenseBuf: SenseBuffer,
}

impl fmt::Debug for SCSI_PASS_THROUGH_DIRECT_WITH_BUFFER {
    fn fmt(&self, fmt: &mut fmt::Formatter) -> fmt::Result {
        fmt::Display::fmt(self, fmt)
    }
}

impl fmt::Display for SCSI_PASS_THROUGH_DIRECT_WITH_BUFFER {
    fn fmt(&self, fmt: &mut fmt::Formatter) -> fmt::Result {
        write!(
            fmt,
            "ScsiPassThrough: ScsiStatus {}, DataTransferLength {}, CDB: opcode {}, flag {}",
            self.sptd.ScsiStatus, self.sptd.DataTransferLength, self.sptd.Cdb[0], self.sptd.Cdb[1],
        )
    }
}

impl SCSI_PASS_THROUGH_DIRECT_WITH_BUFFER {
    pub fn new(dir: u32) -> Self {
        Self {
            sptd: SCSI_PASS_THROUGH_DIRECT {
                Length: size_of::<SCSI_PASS_THROUGH_DIRECT>() as u16,
                ScsiStatus: 0,
                PathId: 0,
                TargetId: 0,
                Lun: 0,
                Cdb: [0; 16],
                CdbLength: 0,
                DataIn: dir as u8,
                DataBuffer: null_mut(),
                DataTransferLength: 0,
                SenseInfoOffset: offset_of!(Self, ucSenseBuf) as u32,
                SenseInfoLength: size_of::<SenseBuffer>() as u8,
                TimeOutValue: 10,
            },
            Filler: 0,
            ucSenseBuf: Default::default(),
        }
    }

    pub fn set_buffer(&mut self, dir: u32, data_src: &[u8]) {
        self.sptd.DataIn = dir as u8;
        self.sptd.DataBuffer = data_src.as_ptr() as *mut c_void;
        self.sptd.DataTransferLength = data_src.len() as u32;
    }
}

impl AsRef<[u8]> for ScsiDataBuffer {
    fn as_ref(&self) -> &[u8] {
        &self.0
    }
}

impl Index<usize> for ScsiDataBuffer {
    type Output = u8;

    fn index(&self, index: usize) -> &Self::Output {
        &self.0[index]
    }
}

impl IndexMut<usize> for ScsiDataBuffer {
    fn index_mut(&mut self, index: usize) -> &mut Self::Output {
        &mut self.0[index]
    }
}

#[repr(C)]
#[allow(non_camel_case_types)]
pub enum ScsiOpcode {
    SCSI_OPCODE_READ_10 = 0x28,
    SCSI_OPCODE_READ_16 = 0x88,
    SCSI_OPCODE_WRITE_10 = 0x2a,
    SCSI_OPCODE_WRITE_16 = 0x8a,
    SCSI_OPCODE_READ_CAPACITY_10 = 0x25,
    SCSI_OPCODE_SERVICE_ACTION_IN = 0x9e,
    SCSI_SERVICE_ACTION_READ_CAPACITY_16 = 0x10,
    SCSI_OPCODE_TEST_UNIT_READY = 0x00,
    SCSI_OPCODE_SECURITY_RECV = 0xa2,
    SCSI_OPCODE_SECURITY_SEND = 0xb5,
}

#[repr(C)]
#[allow(non_camel_case_types)]
pub enum ScsiCdbFlag {
    SCSI_FL_FUA_NV = 0x02,
    SCSI_FL_FUA = 0x08,
    SCSI_FL_DPO = 0x10,
}

#[repr(C)]
#[derive(Debug, Default, Clone, Copy, PackedSize, EncodeBE, DecodeBE)]
pub struct ScsiRwCdb16 {
    pub opcode: u8,
    pub flags: u8,
    pub lba: u64,
    pub len: u32,
    pub group: u8,
    pub control: u8,
}

impl ScsiRwCdb16 {
    pub fn new(opcode: ScsiOpcode, lba: u64, len: u32, flags: u8) -> Self {
        Self {
            opcode: opcode as u8,
            lba,
            len,
            flags,
            ..Default::default()
        }
    }
}

#[repr(C)]
#[derive(Debug, Default, Clone, Copy, PackedSize, EncodeBE, DecodeBE)]
pub struct ScsiSecCdb12 {
    pub opcode: u8,
    pub protocol: u8,
    pub com_id: u16,
    pub reserved: u16,
    pub len: u32,
    pub reserved2: u8,
    pub control: u8,
}

impl ScsiSecCdb12 {
    pub fn new(opcode: ScsiOpcode, protocol: u8, com_id: u16, len: u32) -> Self {
        Self {
            opcode: opcode as u8,
            protocol,
            com_id,
            len,
            ..Default::default()
        }
    }
}
